### PR TITLE
[GlobalISel] Combine [S,U]SUBO

### PR DESCRIPTION
**Pull request by tschuett**
We import the llvm.ssub.with.overflow.* Intrinsics, but the Legalizer also builds them while legalizing other opcodes, see narrowScalarAddSub.

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperArtifacts.cpp [hunk 1]
```diff
@@ -84,3 +84,75 @@ bool CombinerHelper::matchMergeXAndZero(const MachineInstr &MI,
...
+  if (!isLegalOrBeforeLegalizer({TargetOpcode::G_SUB, {DstTy}}) ||
+      !isConstantLegalOrBeforeLegalizer(CarryTy))
+    return false;
+
+  if (Subo->isSigned()) {
+    // G_SSUBO
+    ConstantRange KBLHS = ConstantRange::fromKnownBits(KB->getKnownBits(LHS),
+                                                       /* IsSigned= */ true);
+    ConstantRange KBRHS = ConstantRange::fromKnownBits(KB->getKnownBits(RHS),
+                                                       /* IsSigned= */ true);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperArtifacts.cpp:109]:
```suggestion
                                                       /*IsSigned=*/ true);
    ConstantRange KBRHS = ConstantRange::fromKnownBits(KB->getKnownBits(RHS),
                                                       /*IsSigned=*/ true);
```

I think clang-format wants these to be spaceless

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperArtifacts.cpp [hunk 2]
```diff
@@ -84,3 +84,75 @@ bool CombinerHelper::matchMergeXAndZero(const MachineInstr &MI,
...
+        B.buildSub(Dst, LHS, RHS, MachineInstr::MIFlag::NoSWrap);
+        B.buildConstant(Carry, 0);
+      };
+      return true;
+    }
+    case ConstantRange::OverflowResult::AlwaysOverflowsLow:
+    case ConstantRange::OverflowResult::AlwaysOverflowsHigh: {
+      MatchInfo = [=](MachineIRBuilder &B) {
+        B.buildSub(Dst, LHS, RHS);
+        B.buildConstant(Carry, 1);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperArtifacts.cpp:124]: This needs to consider the target BooleanContents

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperArtifacts.cpp [hunk 3]
```diff
@@ -84,3 +84,75 @@ bool CombinerHelper::matchMergeXAndZero(const MachineInstr &MI,
...
+      B.buildSub(Dst, LHS, RHS, MachineInstr::MIFlag::NoUWrap);
+      B.buildConstant(Carry, 0);
+    };
+    return true;
+  }
+  case ConstantRange::OverflowResult::AlwaysOverflowsLow:
+  case ConstantRange::OverflowResult::AlwaysOverflowsHigh: {
+    MatchInfo = [=](MachineIRBuilder &B) {
+      B.buildSub(Dst, LHS, RHS);
+      B.buildConstant(Carry, 1);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperArtifacts.cpp:151]: Target boolean content

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-overflow.mir [hunk 4]
```diff
@@ -176,3 +176,49 @@ body:             |
...
+    ; CHECK-NEXT: $w1 = COPY %o_wide(s32)
+    ; CHECK-NEXT: RET_ReallyLR implicit $w0
+    %0:_(s32) = COPY $w0
+    %const:_(s32) = G_CONSTANT i32 512
+    %sub:_(s32), %o:_(s1) = G_USUBO %0, %const
+    %o_wide:_(s32) = G_ZEXT %o(s1)
+    $w0 = COPY %sub(s32)
+    $w1 = COPY %o_wide
+    RET_ReallyLR implicit $w0
+...
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-overflow.mir:224]: Add a test with a non-s1 carry output, and vectors

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperArtifacts.cpp [hunk 5]
```diff
@@ -84,3 +84,75 @@ bool CombinerHelper::matchMergeXAndZero(const MachineInstr &MI,
...
+    }
+    }
+    return false;
+  }
+
+  // G_USUBO
+  ConstantRange KBLHS = ConstantRange::fromKnownBits(KB->getKnownBits(LHS),
+                                                     /* IsSigned= */ false);
+  ConstantRange KBRHS = ConstantRange::fromKnownBits(KB->getKnownBits(RHS),
+                                                     /* IsSigned= */ false);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperArtifacts.cpp:136]: Can avoid duplicating these in the signed / unsigned cases

File: llvm/include/llvm/Target/GlobalISel/Combine.td [hunk 6]
```diff
@@ -1901,6 +1907,13 @@ def cmp_combines: GICombineGroup<[
   redundant_binop_in_equality
 ]>;
 
+
+def artifact_combines: GICombineGroup<[
+  merge_combines,
+  match_addos,
+  match_subos
```
Comment (arsenm) [inline llvm/include/llvm/Target/GlobalISel/Combine.td:1914]: These names aren't great. match_subo_no_overflow?

File: llvm/include/llvm/Target/GlobalISel/Combine.td [hunk 7]
```diff
@@ -1901,6 +1907,13 @@ def cmp_combines: GICombineGroup<[
   redundant_binop_in_equality
 ]>;
 
+
+def artifact_combines: GICombineGroup<[
+  merge_combines,
+  match_addos,
```
Comment (aemerson) [inline llvm/include/llvm/Target/GlobalISel/Combine.td:1913]: These aren't artifacts.

Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1913]: Well, the legalizer builds addos while narrowing, e.g., adds:
https://github.com/llvm/llvm-project/blob/764275949897533a4be0728250e69a94d228fbc5/llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp#L6289

Maybe artifacts is too strict. I stated in the summary that we import addos and the legalizer builds while legalizing non addos.

Comment (aemerson) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1913]: That doesn't make them artifacts. There are lots of instructions that the legalized builds like loads/stores but those aren't artifacts either.

Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1913]: Ok, I need a mechanism to combine addos in AArch64PostLegalizerCombiner and I want to have a group where I/we can add more combines that run in the post legalizer  combiner. It is sufficient to rename `artifact_combines` to something like `post_legalizer_combines` ?

Comment (aemerson) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1913]: Just create a new group for these? And then add the group to the post-legalizer combiner set.

Comment (aemerson) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1913]: Shouldn't name these as post-legalizer combines in this file. Each target defines their own pipeline. Name the groups based on their category and then refer to them in any custom groups needed within AArch64Combine.td

Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1912]: I splitted them again into two topic groups.

File: llvm/include/llvm/Target/GlobalISel/Combine.td [hunk 8]
```diff
@@ -1385,6 +1385,12 @@ def match_addos : GICombineRule<
         [{ return Helper.matchAddOverflow(*${root}, ${matchinfo}); }]),
   (apply [{ Helper.applyBuildFn(*${root}, ${matchinfo}); }])>;
 
+def match_subo_no_overflow : GICombineRule<
+  (defs root:$root, build_fn_matchinfo:$matchinfo),
+  (match (wip_match_opcode G_SSUBO, G_USUBO):$root,
```
Comment (aemerson) [inline llvm/include/llvm/Target/GlobalISel/Combine.td:1390]: Could you just use `match` instead of `wip_match_opcode`?

**Review by aemerson — COMMENTED** [review]
One last thing, also not sure if AMDGPU tests are better or not, wait for Matt to finish looking at it.

File: llvm/include/llvm/Target/GlobalISel/Combine.td (hunk 8 above)
Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1390]: Firstly, it was copy-pasted from `match_addos` above. Secondly, I believe we can match only one tree/graph per match. I can split it into two combines if you want.

Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1390]: I don't like it either. There is no prior-art for matches with two output operands.

Comment (aemerson) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1390]: It's not a big deal, we can try to covert them en-masse in another change.

Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1390]: We real/only issue with `wip_match_opcode`is when the docstring says, e.g.  G_FNEG(G_BUILD_VECTOR), and the user is doing `wip_match_opcode` with G_FNEG. It looses precision.

**Review by arsenm — APPROVED** [review]

**Comment by aemerson** [comment]
I have another request in general for your changes. Could you please add some more appropriate descriptions to your PRs so that when you merge them, we end up with a better commit message?

For example, for this commit, we could've done with a description of what the actual combine did. Instead, we got some commentary on some implementation details that aren't very helpful to the reader of the commit.

**Pull request merged**